### ISSUE TITLE
More info on context manager/`strategy.scope` in Custom training loops with tf.distribute.MirroredStrategy()

### ISSUE THREAD

**Issue by 8bitmp3** [tensorflow/tensorflow#41376]
Hi @lamberta and team,

Just some ideas for the **Custom training loops** tutorial (with `tf.distribute.MirroredStrategy()`):

**URLs:** 

https://www.tensorflow.org/tutorials/distribute/custom_training#define_the_loss_function
https://www.tensorflow.org/tutorials/distribute/custom_training#define_the_metrics_to_track_loss_and_accuracy
https://www.tensorflow.org/tutorials/distribute/custom_training#create_a_strategy_to_distribute_the_variables_and_the_graph

**Description:**

In the tutorial, you create a new strategy (`strategy = tf.distribute.MirroredStrategy()`), which has a context manager—`strategy.scope`. Similar to the **[Distributed training with Keras](https://www.tensorflow.org/tutorials/distribute/keras#define_distribution_strategy)** tutorial, maybe you can also explicitly mention what the scope does in the **Custom training loops** tutorial, because some users may not have studied that in greater depth. And, it would make the tutorial more consistent with some of the other ones under **Tutorials** > **Advanced** > **Distributed training**. 

So, to ensure the users don't have to use search here, especially if some of them are new to this, it may be useful to provide a bit more background on the context manager.

Note that inside the context manager—`strategy.scope`—we have to define a number of things: the loss function, metrics, the model itself, Adam optimizer, and checkpoint. And the tutorial separates these items into different sections and provides detailed explanations of each step. Therefore, I suggest maybe we add a few comments as follows:

**Examples:**

```python
#  <NEW> Enclose the loss function call inside the context manager.
with strategy.scope():
  # <original comment>
  <define a method for loss computation>
```

```python
# <NEW> test loss and training and test accuracy metrics must be created under `strategy.scope`
with strategy.scope():
  test_loss = tf.keras.metrics.Mean(name='test_loss')
  train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
      name='train_accuracy')
  test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
      name='test_accuracy')
```

Additionally, this note from the tutorial here may also be improved (the diff is _italicized_) as follows:

> **Note:** You can put all the code below inside a single scope _—the `strategy.scope` context manager that instructs how to split the training among devices_. We are dividing it into several code cells for illustration purposes.

Lastly, similar to this [step](https://www.tensorflow.org/tutorials/distribute/keras#define_distribution_strategy), we can add a short decriptive paragraph (_italicized_) to explain what we're doing when defining `strategy` as `tf.distribute.MirroredStrategy()`:

> _Create a `MirroredStrategy` object to handle distribution, and provides a context manager (`tf.distribute.MirroredStrategy.scope`) to build your model inside:_
>
> ```python
> # If the list of devices is not specified in the
> # `tf.distribute.MirroredStrategy` constructor, it will be auto-detected.
> strategy = tf.distribute.MirroredStrategy()
> ```

---

There is another similar example from the **Multi-worker training with Keras** tutorial:

- [Train the model with MultiWorkerMirroredStrategy](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras#train_the_model_with_multiworkermirroredstrategy):

```python
with strategy.scope():
  # Model building/compiling need to be within `strategy.scope()`.
  multi_worker_model = build_and_compile_cnn_model()
```

**PRs:**

Can submit a PR if you OK (some) of the ideas.

Cheers!

**Comment by 8bitmp3** [comment]
@lamberta Let me know if a small PR would make sense.

The diffs:

```diff
+ # Enclose the loss function call inside the context manager.
with strategy.scope():
  # Set reduction to `none` so we can do the reduction afterwards and divide by
  # global batch size.
  loss_object = tf.keras.losses.SparseCategoricalCrossentropy(
      from_logits=True,
      reduction=tf.keras.losses.Reduction.NONE)
  ...
```

```diff
+ # Instantiate the test loss and training and test accuracy metrics inside the context manager.
with strategy.scope():
  test_loss = tf.keras.metrics.Mean(name='test_loss')
  train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
      name='train_accuracy')
  test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
      name='test_accuracy')
```

```diff
⭑ Note: You can put all the code below inside a single scope
+ —the `strategy.scope` context manager that instructs how to split the training among devices. 
We are dividing it into several code cells for illustration purposes.
```

```diff
⭑ Note: You can put all the code below inside a single scope. We are dividing it into several code cells for illustration purposes.

+ Create a `MirroredStrategy` object to handle distribution, and provides a context manager 
+ (`tf.distribute.MirroredStrategy.scope`) to build your model inside:

# If the list of devices is not specified in the
# `tf.distribute.MirroredStrategy` constructor, it will be auto-detected.
strategy = tf.distribute.MirroredStrategy()
```

WDYT? Thanks @lamberta

**Comment by nikitamaia** [comment]
Hi @8bitmp3, sorry for the late response on this. I'm not sure how much (if at all) the custom training loop tutorial has changed since you wrote this issue. But if you have some small changes that you think will make it easier for users to understand, please feel free to file a PR.

**Comment by 8bitmp3** [comment]
@nikitamaia No worries, will do. Thank you 👍
